---
layout: post
title: 예외처리
categories: JAVA
tags: [JAVA]
---
## 오류
프로그램을 실행하다 보면 특정 원인으로 비정상적인 동작을 일으키며 프로그램이 종료되는 현상이 있는데 이때 우리는 프로그램에 오류가 발생했다고 한다

오류의 종류는 우리가 컴파일할때 발생할 수 있는 컴파일 오류와 실행 중 발생되는 런타임 오류 두 종류가 있다. 컴파일 오류는 잡기 쉽지만, 런타임 오류는 잡기가 까다롭다.

자바에서는 런타임 오류를 에러와 예외 두 종류로 보고 있다.

에러는 프로그램이 코드로 복구될 수 없는 오류를 의미하고 예외는 프로그래머가 직접 예측하여 막을 수 있는 처리가능한 오류라고 보면 된다.

보통 컴파일 시 발생하는 예외의 경우 IDE 사용 시 예외를 처리하라고 표시한다.

하지만 컴파일 시 발견하지 못하는 예외의 경우 런타임 에러라고 하는데 이때는 프로그래머가 직접 예측하여 처리해야 한다.

그리고 그런 예외가 발생했을때 어떤 동작을 처리해야하는지를 예외 처리라고 한다.

오류 예시(어떤 수를 0으로 나눌 경우)
<pre>
<code>
public class ErrorTest {
    public static void main(String[] args) {
        int a,b;
        a=10;
        b=0;

        int c=a/b;
        System.out.println(c);
    }
}
</code>
</pre>
출력 :
<pre>
<code>
Exception in thread "main" java.lang.ArithmeticException: / by zero
	at test.ErrorTest.main(ErrorTest.java:9)
</code>
</pre>

## try, catch
예외가 발생했을 때 이를 처리하거나 메소드를 호출한 곳으로 던진다

*모든 예외는 Exception이라는 클래스를 상속받는다

예시 :
<pre>
<code>
public static void main(String[] ar){
	int a,b;
	a=10;
	b=0;
	try {
		int c=a/b;
		System.out.println(c);	//예외발생으로 실행 불가한 코드
	}catch(ArithmeticException e) { //ArithmeticException이 발생했다면
		System.out.println("ArithmeticException 발생");
		System.out.println("0으로 나눌 수는 없습니다");
		e.printStackTrace(); //어느 부분에서 예외가 발생했는지 알려주는 추적로그
	}finally { //어떤 예외가 발생하던 말던 무조건 실행
		System.out.println("finally 실행");
	}
}
</code>
</pre>
출력 :
<pre>
<code>
ArithmeticException 발생
0으로 나눌 수는 없습니다
Exception in thread "main" java.lang.ArithmeticException: / by zero
	at test.ErrorTest.main(ErrorTest.java:9)
finally 실행
</code>
</pre>
## throws
메소드를 호출한 곳으로 예외 처리를 떠넘기는 것

예시 :
<pre>
<code>
public class ErrorTest {
    public static void divide(int a,int b) throws ArithmeticException {
        if(b==0) throw new ArithmeticException("안 돼"); //devide()를 호출한 main에서 처리
        int c=a/b;
        System.out.println(c);
    }
    public static void main(String[] ar){
        int a=10;
        int b=0;

        try {
            divide(a,b);
        }catch(ArithmeticException e) {
            e.getMessage(); //Exception 생성자 중에서 메시지를 받는 생성자가 있는데, 메시지를 볼 때 이용
            e.printStackTrace();
        }
    }
}
</code>
</pre>
출력 :
<pre>
<code>
java.lang.ArithmeticException: 안 돼
	at test.ErrorTest.divide(ErrorTest.java:5)
	at test.ErrorTest.main(ErrorTest.java:14)
</code>
</pre>
## 예외

자바의 예외에는 CheckedException과 UncheckedException이 있다

|   /   | CheckedException | UncheckedEcxeption |
| ---- | --------------- | ----------------- |
| 처리여부 | 반드시 예외를 처리해야 함 | 명시적인 처리를 강제하지 않음 | 
| 확인시점 | 컴파일 단계 | 실행단계 |
| 예외발생시 트랜잭션 처리 | 롤백하지 않음 | 롤백함 |
| 대표 예외 | Exception의 상속받는 하위 클래스 중<br>RuntimeException을 제외한 모든 예외<br> 1. IOException 2. SQLException| RuntimeException 하위 예외<br><ul><li>NullPointerException</li><li>IllegalArgumentException</li><li>IndexOutOfBoundException|

|:     Fruits \|\| Food           :|||
|:-------- |:-------- |:------------ |
| Apple    |: Apple  :|    Apple     \
| Banana   |  Banana  |    Banana    \
| Orange   |  Orange  |    Orange    |
|:   Rowspan is 5   :||:  How's it? :|
|^^   A. Peach       ||^^ 1. Fine    |
|^^   B. Orange      ||^^ 2. Bad  $I = \int \rho R^{2} dV$     |
|^^   C. Banana      ||   It's OK! ![example image][my-image]  |


CheckedException과 UncheckedException의 가장 명확한 구분 기준은 꼭 처리를 해야 하느냐다. CheckedException이 발생할 가능성이 있는 메소드라면 반드시 로직을 try/catch나 throw를 써야 한다. 반면 UncheckedException의 경우 명시적인 예외처리를 하지 않아도 된다. 이 예외는 피할 수 있지만 개발자가 부주의해서 발생하는 경우가 대부분이고, 미리 예측하지 못했던 상황에서 발생하는 예외가 아니기 때문에 굳이 로직으로 처리할 필요가 없도록 만들어져 있다

또한 예외를 확인할 수 있는 시점에서도 구분할 수 있는데 일반적으로 컴파일 단계에서 명확하게 Exception 체크가 가능한 것을 CheckedException이라 하고, 실행 과정 중 어떠한 특정 논리에 의해 발견되는 Exception을 UncheckedException이라고 한다. 따라서 컴파일 단계에서 확인할 수 없는 예외라 하여 UncheckedException이고, 실행 과정 중 발견된다고 하여 RuntimeException이라 한다.